lib/apiservers/service/restapi/handlers/vch_put.go

```diff
@@ -0,0 +1,339 @@
+		}
+
+		c.DisplayName = vch.Name
+
+		// TODO [AngieCris]: prototype stage, only debug configurable
+		debug := int(vch.Debug)
+		c.Debug.Debug = &debug
+
+		// TODO [AngieCris]: timeout is not configurable from API. Make it less hacky
+		// Set default timeout to 3 minutes
+		c.Timeout = DefaultTimeout
+	}
+
+	return c, nil
+}
+
+func (h *vchPut) getOldVCHConfigAndData(op trace.Operation, d *data.Data, hc *client.HandlerClient) (*config.VirtualContainerHostConfigSpec, *data.Data, error) {
+	vch, err := hc.GetVCHVM(op, d)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	vchConfig, err := hc.GetSecretConfigForVCH(op, vch)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	oldData, err := hc.GetDataFromVCHConfig(op, vch, vchConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return vchConfig, oldData, nil
+}
+
+func (h *vchPut) mergeData(op trace.Operation, oldData *data.Data, newData *data.Data) (*data.Data, error) {
+	err := h.checkImmutableFieldsIfSet(op, oldData, newData)
+	if err != nil {
+		return nil, err
+	}
+
+	err = oldData.CopyNonEmpty(newData)
+	if err != nil {
+		return nil, errors.NewError(http.StatusInternalServerError, "error copying new VCH data: %s", err)
+	}
+
+	// TODO [AngieCris]: there're also a bunch of other fields not copied by CopyNonEmpty, and treated separately (they're still mutable tho). Check with CLI code and figure out a plan
+
+	return oldData, nil
+}
+
+func (h *vchPut) handleConfigure(op trace.Operation, c *configure.Configure, vch *models.VCH, config *config.VirtualContainerHostConfigSpec, hc *client.HandlerClient) error {
+	validator := hc.Validator()
+
+	grantPerms := false
+	if vch.Endpoint != nil && vch.Endpoint.OperationsCredentials != nil {
+		grantPerms = vch.Endpoint.OperationsCredentials.GrantPermissions
+	}
+
+	newConfig, err := validator.Validate(op, c.Data, false)
+	if err != nil {
+		return errors.NewError(http.StatusInternalServerError, "cannot validate configuration: %s", err)
+	}
+
+	// copy changed config from new to old
+	// TODO: [AngieCris] refactor this code to not depend on CLI (option: maybe move this changeConfig logic to vchConfig package)
+	c.CopyChangedConf(config, newConfig, grantPerms)
+
+	// add deprecated fields
+	// TODO [AngieCris]: there're a bunch of other extra fields set separately
+	vConfig := validator.AddDeprecatedFields(op, config, c.Data)
+	vConfig.Timeout = c.Timeout
+	vConfig.VCHSizeIsSet = c.ResourceLimits.IsSet
+
+	// TODO [AngieCris]: handle affinity VM group thing separately
+
+	// TODO [AngieCris]: set UpgradeInProgress flag, and RollBack logic
+
+	err = hc.Executor().Configure(config, vConfig)
+	if err != nil {
+		return errors.NewError(http.StatusInternalServerError, "failed to configure VCH: %s", err)
+	}
+
+	// TODO [AngieCris]: Question: what happens if configure failed and left a broken VCH? Is it the API's job to manually rollback to the original state? (or that's what rollback does)
```

Comment:
I think we should roll back when a failure occurs, like the CLI does.

Determining _when_ to roll back is the tricky part though. The CLI uses a timeout and just sort of assumes that if it hasn't successfully completed by that timeout then something has gone wrong, but that's not a very good pattern.